File: fixtures/jwt-auth/sqlalchemy/two_factor_setting.py
from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from typing import Optional
from datetime import datetime

from .base import Base


class TwoFactorSetting(Base):
    __tablename__ = "two_factor_setting"

    id: Mapped[str] = mapped_column(String, primary_key=True)

    user_id: Mapped[str] = mapped_column(String, ForeignKey("user.id"), unique=True, nullable=False)

    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    method: Mapped[Optional[str]] = mapped_column(Enum("TOTP", "SMS", "EMAIL", name="two_factor_method"), nullable=True)

    totp_secret: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    phone_number: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    backup_codes: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    verified_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, server_default=func.now())

    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, onupdate=func.now())

    user: Mapped["User"] = relationship("User", back_populates="two_factor_setting")

File: fixtures/jwt-auth-pg/sea_orm/two_factor_setting.rs
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "two_factor_settings")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    #[sea_orm(column_type = "Uuid")]
    pub id: String,
    #[sea_orm(unique)]
    #[sea_orm(column_type = "Uuid")]
    pub user_id: String,
    #[sea_orm(default_value = false)]
    pub enabled: bool,
    pub method: Option<TwoFactorMethod>,
    pub totp_secret: Option<String>,
    #[sea_orm(column_type = "String(StringLen::N(20))")]
    pub phone_number: Option<String>,
    pub backup_codes: Option<String>,
    #[sea_orm(column_type = "TimestampWithTimeZone")]
    pub verified_at: Option<DateTimeUtc>,
    #[sea_orm(column_type = "TimestampWithTimeZone")]
    pub created_at: DateTimeUtc,
    #[sea_orm(column_type = "TimestampWithTimeZone")]
    pub updated_at: DateTimeUtc,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::UserId",
        to = "super::user::Column::Id"
    )]
    User,
}

impl Related<super::user::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}